src/ess/reflectometry/tools.py

```diff
@@ -108,3 +112,160 @@ def linlogspace(
         grids.append(mesh[dim, start:])
 
     return sc.concat(grids, dim)
+
+
+def _sort_by(a, by):
+    return [x for x, _ in sorted(zip(a, by, strict=True), key=lambda x: x[1])]
+
+
+def _find_interval_overlaps(intervals):
+    '''Returns the intervals where at least
+    two or more of the provided intervals
+    are overlapping.'''
+    edges = list(chain.from_iterable(intervals))
+    is_start_edge = list(chain.from_iterable((True, False) for _ in intervals))
+    edges_sorted = sorted(edges)
+    is_start_edge_sorted = _sort_by(is_start_edge, edges)
+
+    number_overlapping = 0
+    overlap_intervals = []
+    for x, is_start in zip(edges_sorted, is_start_edge_sorted, strict=True):
+        if number_overlapping == 1 and is_start:
+            start = x
+        if number_overlapping == 2 and not is_start:
+            overlap_intervals.append((start, x))
+        if is_start:
+            number_overlapping += 1
+        else:
+            number_overlapping -= 1
+    return overlap_intervals
+
+
+def _searchsorted(a, v):
+    for i, e in enumerate(a):
+        if e > v:
+            return i
+    return len(a)
+
+
+def _create_qgrid_where_overlapping(qgrids):
+    '''Given a number of Q-grids, construct a new grid
```

Comment:
An alternative could be to just make a simple linspace grid that covers from min to max of the entire range?
I would simplify the logic/code, but maybe the changing resolution q-grid was required by the scientists?

Reply:
Yes that is an alternative. But, 
* how dense do you make the grid?
    * The Q-grid density might differ over the Q-range (the grid might be geom-space),
    * and it might differ between the different curves.
    * Large parts of the Q-range might not have overlapping curves.
    * Taking all those things into consideration you might have to create a linspace grid that is everywhere as dense as the densest region of the densest Q-grid, to be sure you are not neglecting important features of the reflectivity curves. That would be wasteful.